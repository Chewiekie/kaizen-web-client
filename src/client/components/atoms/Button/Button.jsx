/* eslint-disable react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';

import './Button.scss';

const colors = ['primary', 'secondary', 'warning', 'light'];
const iconModes = ['circle-0', 'circle', '1', '2', '3', '4'];

const Button = (props) => {
  const {
    onClick,
    children,
    type,
    icon,
    iconMode,
    iconRight,
    centered,
    color,
    className,
    disabled,
    form,
  } = props;

  const isIcon = type === 'icon';

  const setClass = (isIcon, array, value, index) => isIcon && (array[index] === value);

  const buttonClassName = clsx({
    '--spaced': !isIcon,
    'btn': !isIcon,
    'btn--primary': setClass(!isIcon, colors, color, 0),
    'btn--secondary': setClass(!isIcon, colors, color, 1),
    'btn--warning': setClass(!isIcon, colors, color, 2),
    'btn-icon': isIcon,
    'btn-icon--primary': setClass(isIcon, colors, color, 0),
    'btn-icon--secondary': setClass(isIcon, colors, color, 1),
    'btn-icon--warning': setClass(isIcon, colors, color, 2),
    'btn-icon--light': setClass(isIcon, colors, color, 3),
    'btn-icon--space-mode-circle-0': setClass(isIcon, iconModes, iconMode, 0),
    'btn-icon--space-mode-circle': setClass(isIcon, iconModes, iconMode, 1),
    'btn-icon--space-mode-1': setClass(isIcon, iconModes, iconMode, 2),
    'btn-icon--space-mode-2': setClass(isIcon, iconModes, iconMode, 3),
    'btn-icon--space-mode-3': setClass(isIcon, iconModes, iconMode, 4),
    'btn-icon--space-mode-4': setClass(isIcon, iconModes, iconMode, 5),
    [className]: className,
  });

  const contentClassName = clsx({
    'btn__content': true,
    'btn__content--cols2': !isIcon && (icon && centered),
    'btn__content--cols3': !isIcon && (icon && !centered && !iconRight),
    'btn__content--cols3-right': !isIcon && (icon && !centered && iconRight),
  });

  return (
    <button
      onClick={onClick}
      type={type}
      className={buttonClassName}
      disabled={disabled}
      form={form}
    >
      {icon ? (
        <div className={contentClassName}>
          {!iconRight && <div className='btn__container--icon'>{icon}</div>}
          <div className='btn__container--children'>{children}</div>
          {iconRight && <div className='btn__container--icon'>{icon}</div>}
        </div>
      ) : children}
    </button>
  );
};

Button.propTypes = {
  /** Function that will be called on click event. */
  onClick: PropTypes.func.isRequired,
  /** The text inside the button or another component */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /** Specify the type of the button */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  /** Element Icon type */
  icon: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  /** Spacing modes for icon type */
  iconMode: PropTypes.oneOf(iconModes),
  /** Indicates is placed on the right side of the text */
  iconRight: PropTypes.bool,
  /** Icon and other elements will be centered. */
  centered: PropTypes.bool,
  /** Boolean class to specify if the color button is primary, secondary or warning */
  color: PropTypes.oneOf(colors),
  /** Class to overwrite the styles */
  className: PropTypes.string,
  /** Specify if the button is disabled or not */
  disabled: PropTypes.bool,
  /** If the button is outside a form and you want to active the submit event on click, you can pass the form id */
  form: PropTypes.string,
};

Button.defaultProps = {
  type: 'button',
  icon: null,
  iconMode: null,
  iconRight: false,
  centered: false,
  color: 'primary',
  className: '',
  disabled: false,
  form: '',
};

export default Button;
