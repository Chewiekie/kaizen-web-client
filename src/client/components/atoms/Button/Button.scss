@mixin btnActive($color: initial, $colorBg: initial) {
	&:active:hover {
		box-shadow: inset 0 0 6px 2px $color;
		@include colorFontAndBg($color, transparent);
	}
}

@mixin btnHover($color: initial, $colorBg: initial) {
	&:hover {
		@include colorFontAndBg($color, $colorBg);
	}
}

.btn {
	display: inline-block;
	border: none;
	width: auto;
	font-size: 1rem;
	font-weight: $fontWeightBold;
	outline: none;
	cursor: pointer;
	border-radius: getSpacing(3);
	@include colorFontAndBg($colorFontDark, transparent);
	@include btnHover($colorFontLight, $colorPrimaryMain);
  @include btnActive($colorSuccess, $colorFontDark);

  &__container {
    &--icon {
      display: flex;
    }
  }

	&--primary {
		@include colorFontAndBg($colorFontLight, $colorPrimaryMain);
		@include btnHover($colorFontDark, $colorSecondaryMain);
	}

	&--secondary {
		@include colorFontAndBg($colorFontDark, $colorSecondaryMain);
		@include btnHover($colorFontLight, $colorPrimaryMain);
	}

	&--warning {
    @include colorFontAndBg($colorFontLight, $colorWarning);
		@include btnHover($colorFontDark, $colorError);
  }

  &--spaced {
    padding: getSpacing(1);
    @include applyTo('m') {
      padding: getSpacing(1) getSpacing(4);
    }
  }

	&:disabled {
		@include colorFontAndBg($colorFontDisabled, $colorBgDisabled);
		cursor: not-allowed;
  }

  &__content {
    &--cols3 {
      display: grid;
      grid-template-columns: 1fr auto 1fr;
      align-items: center;
      column-gap: getSpacing(1);
    }

    &--cols3-right {
      @extend .btn__content--cols3;
      &>:nth-child(1) {
        grid-column: 2 / 3;
      }
      &>:nth-child(2) {
        grid-column: 3 / 4;
      }
    }

    &--cols2 {
      @extend .btn__content--cols3;
      grid-template-columns: auto auto;
    }
  }
}

.btn-icon {
	@extend .btn;
	$padTB: getSpacing(0.5);
	$padRL: getSpacing(1);
	line-height: 0;
	width: fit-content;
	height: fit-content;
	border-radius: getSpacing(0);
	@include btnHover($colorPrimaryMain, transparent);
	@include btnActive(initial, initial);

	&--primary {
		@include colorFontAndBg($colorPrimaryMain, transparent);
		@include btnHover($colorSecondaryMain, transparent);
	}

	&--secondary {
		@include colorFontAndBg($colorSecondaryMain, transparent);
		@include btnHover($colorPrimaryMain, transparent);
	}

	&--warning {
		@include colorFontAndBg($colorWarning, transparent);
		@include btnHover($colorError, transparent);
	}

	&--light {
		@include colorFontAndBg($colorFontLight, transparent);
		@include btnHover($colorPrimaryMain, transparent);
	}

	&:disabled {
		@include colorFontAndBg($colorFontDisabled, $colorBgDisabled);
		cursor: not-allowed;
  }

	&--space-mode-circle-0 {
		border-radius:100%;
	}

	&--space-mode-circle {
		@include padding($padRL, $padRL, $padRL, $padRL);
		border-radius:100%;
	}

	&--space-mode-1 {
		@include padding($padTB, $padRL, $padTB, $padRL);
		border-radius: getSpacing(0.5);
	}

	&--space-mode-2 {
		@include padding($padTB, $padRL, $padTB, $padRL);
		border-radius: getSpacing(1);
	}

	&--space-mode-3 {
		@include padding($padTB, $padRL, $padTB, $padRL);
		border-radius: getSpacing(2);
	}

	&--space-mode-4 {
		@include padding($padTB, $padRL, $padTB, $padRL);
		border-radius: getSpacing(4);
	}
}
