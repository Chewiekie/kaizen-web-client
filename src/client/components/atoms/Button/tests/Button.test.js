import React from 'react';
import { create } from 'react-test-renderer';
import { mount, shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Button from '../Button';

configure({ adapter: new Adapter() });

describe('Button', () => {
  test('Match Snapshot', () => {
    const button = create(<Button onClick={() => null}>Snapshot</Button>);
    expect(button.toJSON()).toMatchSnapshot();
  });

  test('Button have class .btn', () => {
    const button = shallow(<Button onClick={() => null}>Btn</Button>);
    expect(button.hasClass('btn')).toBe(true);
  });

  test('Button have class .btn--primary', () => {
    const button = shallow(<Button onClick={() => null}>Btn</Button>);
    expect(button.hasClass('btn--primary')).toBe(true);
  });

  test('Button have class .btn--secondary', () => {
    const button = shallow(<Button secondary onClick={() => null}>Btn</Button>);
    expect(button.hasClass('btn--secondary')).toBe(true);
  });

  test('Button prop type default', () => {
    const button = shallow(<Button onClick={() => null}>Btn</Button>);
    expect(button.props().type).toBe('button');
  });

  test('Button prop type', () => {
    const button = shallow(<Button type='submit' onClick={() => null}>Btn</Button>);
    expect(button.props().type).toBe('submit');
  });

  test('Button prop onClick', () => {
    const onClick = jest.fn();
    const button = shallow(<Button onClick={onClick}>Btn</Button>);
    expect(button.props().onClick).toBe(onClick);
    button.simulate('click');
    expect(onClick.mock.calls.length).toEqual(1);
  });

  test('Button prop className', () => {
    const button = shallow(<Button className='btn--test' onClick={() => null}>Btn</Button>);
    expect(button.hasClass('btn--test')).toBe(true);
  });

  test('Button prop form', () => {
    const button = shallow(<Button form='test-form' onClick={() => null}>Btn</Button>);
    expect(button.props().form).toBe('test-form');
  });

  test('Button prop disabled', () => {
    const onClick = jest.fn();
    const button = mount(<Button disabled onClick={onClick}>Btn</Button>);
    button.simulate('click');
    expect(onClick.mock.calls.length).toEqual(0);
  });

  test('Button prop children', () => {
    const button = shallow(<Button onClick={() => null}>Btn</Button>);
    expect(button.props().children).toBe('Btn');
  });

});
