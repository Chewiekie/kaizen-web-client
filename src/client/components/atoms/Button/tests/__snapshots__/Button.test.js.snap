// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Match Snapshot 1`] = `
<button
  className="btn btn--primary"
  disabled={false}
  form=""
  onClick={[Function]}
  type="button"
>
  Snapshot
</button>
`;
